### PR TITLE
Enhance RestLayerPrivilegesEvaluator Code Coverage 

### PR DESCRIPTION
**Pull request by scrawfor99**
### Description
`RestLayerPrivilegesEvaluator` is expected to have 100% code coverage. This change should boost the coverage as high as is reasonable. Per this comment: https://github.com/opensearch-project/security/issues/2929#issuecomment-1686346021, I am not adding tests for debug logs. 

### Issues Resolved
Resolves https://github.com/opensearch-project/security/issues/2929

### Check List
- [X] New functionality includes testing
- [ ] ~New functionality has been documented~
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by willyborankin — DISMISSED** [review]

**Comment by scrawfor99** [comment]
Screenshot 2023-08-21 at 10 13 24 AM


This makes very little sense in my opinion. I see no reason that the changes, I made would decrease coverage at all.

File: src/test/java/org/opensearch/security/filter/RestPathMatchesTest.java
Comment (peternied) [inline src/test/java/org/opensearch/security/filter/RestPathMatchesTest.java:1]: If code cov says that there is a no change, are you sure these tests are executed and their coverage data submitted?  Do you see these tests executed from the artifacts of the [check runs](https://github.com/opensearch-project/security/actions/runs/5927349083)? 

image

Comment (scrawfor99) [reply src/test/java/org/opensearch/security/filter/RestPathMatchesTest.java:1]: It does not look like it ran. I am not sure how to fix that though. I don't think it is something related to this change--I triggered a re-run.

Comment (willyborankin) [reply src/test/java/org/opensearch/security/filter/RestPathMatchesTest.java:1]: @scrawfor99  I think you need to call it `*Test**s**`.

Comment (willyborankin) [reply src/test/java/org/opensearch/security/filter/RestPathMatchesTest.java:1]: Maybe it is not only your case but others as well.

Comment (peternied) [reply src/test/java/org/opensearch/security/filter/RestPathMatchesTest.java:1]: Nice catch @willyborankin .  I've got a change that will expect 100% code coverage in the tests (e.g. all test code should be executed) that should help us catch this kind of issue

- https://github.com/opensearch-project/security/pull/3196

Comment (willyborankin) [reply src/test/java/org/opensearch/security/filter/RestPathMatchesTest.java:1]: Yes I did the same mistake :-D. Here is PR: https://github.com/opensearch-project/security/pull/3224. Lets see.

Comment (scrawfor99) [reply src/test/java/org/opensearch/security/filter/RestPathMatchesTest.java:1]: Huh, who knew. I will swap over the name.

**Comment by scrawfor99** [comment]
Screenshot 2023-08-22 at 10 13 43 AM


Confirmed tests ran.

**Comment by scrawfor99** [comment]
`RestLayerPrivilegeEvaluator` still lists `isInitialized` as having partial coverage. However, I think this is adequate. I added a test to check that it fails when not initialized (one of the two possibilities) and all other tests which test something else will inherently check `isInitialized` is `true` during their evaluation. This covers the true case.

**Comment by scrawfor99** [comment]
This is the code coverage for the other `SecurityRestFilter`: https://app.codecov.io/gh/opensearch-project/security/pull/3218/blob/src/main/java/org/opensearch/security/filter/SecurityRestFilter.java

It still labels a significant amount as uncovered. I am not sure what is expected here however, as the way to exercise the files s to provide bad inputs and test the handling.

**Review by peternied — COMMENTED** [review]
I've created a pull request for your fork that reduces the product code, adds more assertions, and gets more coverage, take a look and see if you'd like to include it with PR

- https://github.com/scrawfor99/security/pull/4

**Review by DarshitChanpura — DISMISSED** [review]

File: src/test/java/org/opensearch/security/privileges/RestLayerPrivilegesEvaluatorTest.java [hunk 1]
```diff
@@ -95,28 +102,45 @@ public void testEvaluate_Initialized_Success() {
...
+    public void testEvaluate_NotInitialized_NoSecurityRoles_ExceptionThrown() throws Exception {
+        final OpenSearchSecurityException exception = assertThrows(
+            OpenSearchSecurityException.class,
+            () -> privilegesEvaluator.evaluate(TEST_USER, null)
+        );
+        assertThat(exception.getMessage(), equalTo("OpenSearch Security is not initialized."));
+        verify(configModel).getSecurityRoles();
     }
 
     @Test
```
Comment (peternied) [inline src/test/java/org/opensearch/security/privileges/RestLayerPrivilegesEvaluatorTest.java:133]: I think I get you an error in my pull request, let me know if you need a hand addressing it.

```
Suite: Test class org.opensearch.security.privileges.RestLayerPrivilegesEvaluatorTest
  2> org.mockito.exceptions.misusing.InvalidUseOfMatchersException: 
    Invalid use of argument matchers!
    2 matchers expected, 1 recorded:
    -> at org.opensearch.security.privileges.RestLayerPrivilegesEvaluatorTest.testMapRoles_ReturnsMappedRoles(RestLayerPrivilegesEvaluatorTest.java:137)

    This exception may occur if matchers are combined with raw values:
        //incorrect:
        someMethod(any(), "raw String");
    When using matchers, all arguments have to be provided by matchers.
    For example:
        //correct:
        someMethod(any(), eq("String by matcher"));

    For more info see javadoc for Matchers class.
        at app//org.opensearch.security.privileges.RestLayerPrivilegesEvaluatorTest.testMapRoles_ReturnsMappedRoles(RestLayerPrivilegesEvaluatorTest.java:137)

Tests with failures:
 - org.opensearch.security.privileges.RestLayerPrivilegesEvaluatorTest.testMapRoles_ReturnsMappedRoles
```

Comment (scrawfor99) [reply src/test/java/org/opensearch/security/privileges/RestLayerPrivilegesEvaluatorTest.java:133]: No worries, thank you for your help. I will take a look at the failure now.

File: src/test/java/org/opensearch/security/privileges/RestLayerPrivilegesEvaluatorTest.java [hunk 2]
```diff
@@ -13,76 +13,82 @@
...
     @Mock
     private ThreadPool threadPool;
     @Mock
-    private AtomicReference<NamedXContentRegistry> namedXContentRegistry;
-    @Mock
     private ConfigModel configModel;
     @Mock
     private DynamicConfigModel dcm;
     @Mock
     private PrivilegesEvaluatorResponse presponse;
```
[the remark this answers is not in the document]
Comment (DarshitChanpura) [reply src/test/java/org/opensearch/security/privileges/RestLayerPrivilegesEvaluatorTest.java:64]: Added a PR: https://github.com/scrawfor99/security/pull/5

**Review by DarshitChanpura — APPROVED** [review]
LGTM. Thanks @scrawfor99 !

File: src/test/java/org/opensearch/security/privileges/RestLayerPrivilegesEvaluatorTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (scrawfor99) [reply src/test/java/org/opensearch/security/privileges/RestLayerPrivilegesEvaluatorTest.java]: Thanks!

**Review by peternied — APPROVED** [review]

**Comment by DarshitChanpura** [comment]
Backported in #2956 as that is the parent change for this PR.

**Pull request merged**